jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java

```diff
@@ -1204,6 +1204,167 @@ else if (ensured > 0)
+                return new ListIterator<>()
+                {
+                    HttpField last;
+
+                    @Override
+                    public boolean hasNext()
+                    {
+                        return i.hasNext();
+                    }
+
+                    @Override
+                    public HttpField next()
+                    {
+                        return last = i.next();
+                    }
+
+                    @Override
+                    public boolean hasPrevious()
+                    {
+                        return i.hasPrevious();
+                    }
+
+                    @Override
+                    public HttpField previous()
+                    {
+                        return last = i.previous();
+                    }
+
+                    @Override
+                    public int nextIndex()
+                    {
+                        return i.nextIndex();
+                    }
+
+                    @Override
+                    public int previousIndex()
+                    {
+                        return i.previousIndex();
+                    }
+
+                    @Override
+                    public void remove()
+                    {
+                        if (last != null && onRemoveField(last))
+                        {
+                            last = null;
+                            i.remove();
+                        }
+                    }
+
+                    @Override
+                    public void set(HttpField field)
+                    {
+                        if (field == null)
+                        {
+                            if (last != null && onRemoveField(last))
+                            {
+                                last = null;
+                                i.remove();
+                            }
+                        }
+                        else
+                        {
+                            if (last != null && onRemoveField(last))
+                            {
+                                field = onAddField(field);
+                                if (field != null)
+                                {
+                                    last = null;
+                                    i.set(field);
+                                }
+                            }
+                        }
+                    }
+
+                    @Override
+                    public void add(HttpField field)
+                    {
+                        if (field != null)
+                        {
+                            field = onAddField(field);
+                            if (field != null)
+                            {
+                                last = null;
+                                i.add(field);
+                            }
+                        }
+                    }
+                };
+            }
+        }
     }
 
     /**
```

Comment:
Also you should fix the `hashCode()` method on `MutableHttpFields`. 
It is iterating over the array length and not `_size` so will throw NPE if called.